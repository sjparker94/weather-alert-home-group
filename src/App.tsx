import React from 'react';
import '@csstools/normalize.css';
import { BrowserRouter as Router, Route } from 'react-router-dom';

import Pages from './pages/Pages';
import PageLayout from './components/PageLayout/PageLayout';

const App: React.FC = () => {
    // const isInitialLoad = useSelector<AppState, boolean>(state => state.locations.isInitialLoad);
    // const app = !isInitialLoad ? <Route component={Pages} /> : null;
    return (
        <Router>
            <PageLayout>
                <Route component={Pages} />
            </PageLayout>
        </Router>
    );
};

export default App;
