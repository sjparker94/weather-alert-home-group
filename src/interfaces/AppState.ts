import LocationsState from './LocationsState';
import SettingsState from './SettingsState';

interface AppState {
    locations: LocationsState;
    settings: SettingsState;
}

export default AppState;
