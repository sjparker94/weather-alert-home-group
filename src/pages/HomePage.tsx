import React from 'react';
import HomeContentWrapper from '../components/ContentBlock/ContentWrapper';
import SearchForm from '../components/SearchForm/SearchForm';
import ContentBlock from '../components/ContentBlock/ContentBlock';

const HomePage: React.FC = () => {
    return (
        <HomeContentWrapper>
            <h1 className="welcome-heading">Welcome!</h1>
            <p className="welcome-paragraph">
                Use the form below to search for a city and add it to your favourites
            </p>
            <ContentBlock>
                <SearchForm />
            </ContentBlock>
        </HomeContentWrapper>
    );
};

export default HomePage;
