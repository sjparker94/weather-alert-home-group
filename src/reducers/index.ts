import { combineReducers } from 'redux';
import locations from './locationsReducer';
import settings from './settingsReducer';

const rootReducer = combineReducers({
    locations,
    settings,
});

export default rootReducer;
