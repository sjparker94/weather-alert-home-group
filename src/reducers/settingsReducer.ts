import { AllSettingsActions } from '../actions/allActions';
import SettingsState from '../interfaces/SettingsState';
import * as settingsActionTypes from '../constants/settingsActions';

export const settingsInitialState: SettingsState = {
    isFahrenheit: false,
    isKm: false,
};

const settings = (
    state: SettingsState = settingsInitialState,
    action: AllSettingsActions
): SettingsState => {
    switch (action.type) {
        case settingsActionTypes.TEMPERATURE_UNIT_TOGGLE:
            return {
                ...state,
                isFahrenheit: !state.isFahrenheit,
            };
        case settingsActionTypes.SPEED_UNIT_TOGGLE:
            return {
                ...state,
                isKm: !state.isKm,
            };
        default:
            return state;
    }
};

export default settings;
