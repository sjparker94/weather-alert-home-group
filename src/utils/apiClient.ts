import axios from 'axios';

const apiClient = axios.create({
    baseURL: process.env.REACT_APP_OPENWEATHER_API_URL,
    responseType: 'json',
    headers: {
        'Content-Type': 'application/json',
        Accept: 'application/json',
    },
});

export default apiClient;
