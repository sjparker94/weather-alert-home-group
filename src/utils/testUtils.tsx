import Location from '../interfaces/Location';
import React from 'react';
import { ThemeProvider } from 'styled-components';
import { render as rtlRender } from '@testing-library/react';
import { createStore, applyMiddleware } from 'redux';
import { Provider } from 'react-redux';
import rootReducer from '../reducers/index';
import thunk from 'redux-thunk';
import theme from '../styles/theme';

export const fakeLocation = (): Location => ({
    coord: {
        lon: -0.13,
        lat: 51.51,
    },
    weather: [
        {
            id: 802,
            main: 'Clouds',
            description: 'scattered clouds',
            icon: '03d',
        },
    ],
    base: 'stations',
    main: {
        temp: 22.85,
        feels_like: 19.87,
        temp_min: 22.22,
        temp_max: 24.44,
        pressure: 1028,
        humidity: 31,
    },
    visibility: 10000,
    wind: {
        speed: 2.6,
        deg: 0,
    },
    clouds: {
        all: 40,
    },
    dt: 1594560101,
    sys: {
        type: 1,
        id: 1414,
        country: 'GB',
        sunrise: 1594526265,
        sunset: 1594584852,
    },
    timezone: 3600,
    id: 2643743,
    name: 'London',
    cod: 200,
});

const render = (ui: any, initialStore = {}, options = {}) => {
    const store = createStore(rootReducer, initialStore, applyMiddleware(thunk));

    const Providers = ({ children }: any) => (
        <Provider store={store}>
            <ThemeProvider theme={theme}>{children}</ThemeProvider>
        </Provider>
    );

    return rtlRender(ui, { wrapper: Providers, ...options });
};
// re-export everything
export * from '@testing-library/react';

// override render method
export { render };
