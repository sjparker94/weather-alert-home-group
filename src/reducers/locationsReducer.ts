import produce, { Draft, castDraft } from 'immer';
import LocationsState from '../interfaces/LocationsState';
import neverReached from '../utils/neverReached';
import { AllActions } from '../actions/allActions';
import * as LocationsActionTypes from '../constants/actions';
import { findByProp } from '../utils/arrayUtils';
import Location from '../interfaces/Location';

// type LocationsActionTypes = any;
export const locationsInitialState: LocationsState = {
    data: [],
    isInitialLoad: true,
    getLocations: {
        isPending: false,
        success: false,
    },
    searchLocation: {
        isPending: false,
        success: false,
        error: null,
        data: null,
    },
};

const locations = produce((draft: Draft<LocationsState>, action: AllActions) => {
    switch (action.type) {
        case LocationsActionTypes.GET_LOCATIONS_REQUEST:
            draft.getLocations.isPending = true;
            break;
        case LocationsActionTypes.GET_LOCATIONS_SUCCESS:
            draft.data = castDraft(action.payload);
            draft.isInitialLoad = false;
            break;
        case LocationsActionTypes.SEARCH_LOCATION_REQUEST:
            draft.searchLocation.isPending = true;
            draft.searchLocation.data = null;
            draft.searchLocation.error = null;
            break;
        case LocationsActionTypes.SEARCH_LOCATION_SUCCESS:
            draft.searchLocation.isPending = false;
            draft.searchLocation.error = null;
            draft.searchLocation.data = action.payload;
            break;
        case LocationsActionTypes.SEARCH_LOCATION_CANCEL:
            draft.searchLocation.data = null;
            draft.searchLocation.error = null;
            break;
        case LocationsActionTypes.SEARCH_LOCATION_FAIL:
            draft.searchLocation.isPending = false;
            draft.searchLocation.error = action.payload;
            break;
        case LocationsActionTypes.ADD_LOCATION:
            // See if the location is already in favourites
            const currentIdx = draft.data.findIndex(findByProp<Location>('id', action.payload.id));

            if (currentIdx === -1) {
                draft.data.push(action.payload);
            } else {
                // Update existing location with the latest data fetched
                draft.data[currentIdx] = action.payload;
            }
            draft.searchLocation.data = null;
            break;
        default:
            neverReached(action);
    }
}, locationsInitialState);

export default locations;
