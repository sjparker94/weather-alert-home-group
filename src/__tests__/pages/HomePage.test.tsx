import React from 'react';
import axios from 'axios';
import userEvent from '@testing-library/user-event';
import '@testing-library/jest-dom/extend-expect';

// @ts-ignore
import { render, screen, fireEvent, waitFor, fakeLocation } from '../../utils/testUtils';
import HomePage from '../../pages/HomePage';
import App from '../../App';
import AppState from '../../interfaces/AppState';
import { locationsInitialState } from '../../reducers/locationsReducer';
import { settingsInitialState } from '../../reducers/settingsReducer';

jest.mock('axios');

const mockLocation = fakeLocation();
const mockLocation2 = fakeLocation({
    id: 123,
    name: 'Newcastle upon Tyne',
});
const mockLocation3 = fakeLocation({
    id: 456,
    name: 'Somewhere else',
});

const mockInitalReduxStateWithData: AppState = {
    locations: {
        ...locationsInitialState,
        isInitialLoad: false,
        data: [mockLocation, mockLocation2],
    },
    settings: {
        ...settingsInitialState,
    },
};

describe('<App />', () => {
    afterEach(() => {
        window.localStorage.removeItem('locations');
    });

    it('renders and matches snapshot', async () => {
        const { container } = render(<App />);
        // Check that the search input is in the dom
        expect(screen.getByTestId('location-search-input')).toBeInTheDocument();
        expect(container).toMatchSnapshot();
    });
    it('handles user input', async () => {
        render(<App />);
        const inputEl = screen.getByTestId('location-search-input');
        expect(inputEl).toBeInTheDocument();

        await userEvent.type(inputEl, 'London');

        expect(inputEl).toHaveAttribute('value', 'London');
    });
    it('displays confirm block after a successful search', async () => {
        const { getByText, queryByTestId } = render(<HomePage />, mockInitalReduxStateWithData);

        const inputEl = screen.getByPlaceholderText(/Newcastle Upon Tyne/i);
        const buttonEl = screen.getByTestId('main-search-button');
        axios.get = jest.fn().mockResolvedValueOnce({ data: mockLocation3 });

        await userEvent.type(inputEl, 'Somewhere else');
        fireEvent.click(buttonEl);

        // await waitForElement(() => getByText('Richard'));
        await waitFor(() => {
            // Check the api has been called
            expect(axios.get).toHaveBeenCalledTimes(1);
        });
        // // Check any errors are not displaying
        expect(queryByTestId('search-error')).not.toBeInTheDocument();
        // // Check that the return from the api is being displayed
        expect(getByText(mockLocation3.name)).toBeInTheDocument();
    });
    it('displays error block after a search failure', async () => {
        render(<App />);
        const inputEl = screen.getByPlaceholderText(/Newcastle Upon Tyne/i);
        const buttonEl = screen.getByTestId('main-search-button');
        axios.get = jest.fn().mockRejectedValue(new Error());

        await userEvent.type(inputEl, 'lssdk');
        fireEvent.click(buttonEl);

        await waitFor(() => {
            // Check that the error is being displayed
            expect(screen.getByTestId('search-error')).toBeInTheDocument();
        });
    });
    it('handles the cancelling of a location search', async () => {
        const { getByPlaceholderText, getByTestId } = render(<HomePage />);
        const inputEl = getByPlaceholderText(/Newcastle Upon Tyne/i);
        const buttonEl = getByTestId('main-search-button');
        axios.get = jest.fn().mockResolvedValue({ data: mockLocation });

        await userEvent.type(inputEl, 'London');
        fireEvent.click(buttonEl);

        const cancelButton = await screen.findByTestId('cancel-search-button');
        fireEvent.click(cancelButton);

        // Check that the confirm block has been removed from the dom
        expect(screen.queryByText(mockLocation.name)).not.toBeInTheDocument();
    });
    it('handles the confirming of a location search and adds to the favourites', async () => {
        const { getByPlaceholderText, getByTestId } = render(<App />);
        const inputEl = getByPlaceholderText(/Newcastle Upon Tyne/i);
        const buttonEl = getByTestId('main-search-button');
        axios.get = jest.fn().mockResolvedValue({ data: mockLocation });

        userEvent.type(inputEl, 'London');
        fireEvent.click(buttonEl);

        const confirmButton = await screen.findByTestId('confirm-search-button');
        fireEvent.click(confirmButton);

        // Get the list of favourites
        const favouriteLocationEls: HTMLElement[] = screen.getAllByTestId('favourite-location');
        // Check that the favourite locations have been updated and now contain an item
        expect(favouriteLocationEls.length).toBe(1);
        // Check that the correct location has been added
        expect(favouriteLocationEls[0]).toHaveTextContent(mockLocation.name);
    });
});
