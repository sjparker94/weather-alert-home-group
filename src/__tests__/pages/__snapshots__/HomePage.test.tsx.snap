// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App /> renders and matches snapshot 1`] = `
<div>
  <div
    class="sc-fznWqX foteFV"
  >
    <header
      class="sc-fznJRM ihGeLz"
    >
      <div
        class="header-logo-wrapper"
      >
        <a
          href="/"
        >
          <span
            class="logo"
          >
            Weather Connect
          </span>
        </a>
      </div>
      <div
        class="header-setting-wrapper"
      >
        <button
          aria-label="Open Settings"
        >
          <svg
            fill="#FFFFFF"
            height="24"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <g>
              <path
                d="M0,0h24v24H0V0z"
                fill="none"
              />
              <path
                d="M19.14,12.94c0.04-0.3,0.06-0.61,0.06-0.94c0-0.32-0.02-0.64-0.07-0.94l2.03-1.58c0.18-0.14,0.23-0.41,0.12-0.61 l-1.92-3.32c-0.12-0.22-0.37-0.29-0.59-0.22l-2.39,0.96c-0.5-0.38-1.03-0.7-1.62-0.94L14.4,2.81c-0.04-0.24-0.24-0.41-0.48-0.41 h-3.84c-0.24,0-0.43,0.17-0.47,0.41L9.25,5.35C8.66,5.59,8.12,5.92,7.63,6.29L5.24,5.33c-0.22-0.08-0.47,0-0.59,0.22L2.74,8.87 C2.62,9.08,2.66,9.34,2.86,9.48l2.03,1.58C4.84,11.36,4.8,11.69,4.8,12s0.02,0.64,0.07,0.94l-2.03,1.58 c-0.18,0.14-0.23,0.41-0.12,0.61l1.92,3.32c0.12,0.22,0.37,0.29,0.59,0.22l2.39-0.96c0.5,0.38,1.03,0.7,1.62,0.94l0.36,2.54 c0.05,0.24,0.24,0.41,0.48,0.41h3.84c0.24,0,0.44-0.17,0.47-0.41l0.36-2.54c0.59-0.24,1.13-0.56,1.62-0.94l2.39,0.96 c0.22,0.08,0.47,0,0.59-0.22l1.92-3.32c0.12-0.22,0.07-0.47-0.12-0.61L19.14,12.94z M12,15.6c-1.98,0-3.6-1.62-3.6-3.6 s1.62-3.6,3.6-3.6s3.6,1.62,3.6,3.6S13.98,15.6,12,15.6z"
              />
            </g>
          </svg>
        </button>
      </div>
    </header>
    <div
      class="main-page-layout-wrapper"
    >
      <main
        class="main-page-content"
      >
        <section
          class="sc-fzozJi eRVneo"
        >
          <h1
            class="welcome-heading"
          >
            Welcome!
          </h1>
          <h3
            class="welcome-paragraph"
          >
            Use the form below to search for a city and add it to your favourites
          </h3>
        </section>
        <div
          class="sc-AxjAm ljWcjv"
        >
          <div
            class="sc-AxheI kwSqZm"
            title="Location Search"
          >
            <div
              class="sc-AxgMl aECWc"
            >
              <h3>
                Location Search
              </h3>
            </div>
            <form
              class="sc-AxhUy hPtbI"
            >
              <div
                class="inputs-wrapper"
              >
                <div
                  class="sc-AxiKw hwsIHP"
                >
                  <label
                    for="city"
                  >
                    Enter the name of a city
                  </label>
                  <input
                    aria-label="Enter the name of a city"
                    data-testid="location-search-input"
                    id="city"
                    name="city"
                    placeholder="Newcastle upon Tyne"
                    type="text"
                    value=""
                  />
                  <p
                    class="input-description"
                  >
                    E.g. Newcastle Upon Tyne or New York
                  </p>
                </div>
              </div>
              <button
                class="sc-AxirZ bIZjPC"
                data-testid="main-search-button"
                disabled=""
              >
                <span
                  class="material-icons"
                >
                  search
                </span>
                Search
              </button>
            </form>
          </div>
        </div>
      </main>
      <div
        class="sc-fzqBZW dgWXPS"
      >
        <div
          class="favourites-wrapper"
        >
          <h2
            class="main-title"
          >
            <span
              class="material-icons star-icon"
            >
              star
            </span>
             Your Favourites
          </h2>
          <div
            class="no-locations-saved"
            data-testid="favourites-no-locations"
          >
            <p>
              No locations have been selected use the form on the home page to add one
            </p>
          </div>
        </div>
        <footer
          class="sc-fzoXzr dXVXzl"
        >
          <p>
            Made with â™¥ by Simon Parker
          </p>
        </footer>
      </div>
    </div>
  </div>
</div>
`;
